torchmetrics/classification/binned_precision_recall.py

```diff
@@ -0,0 +1,303 @@
+
+    - ``target`` (long tensor): ``(N, ...)`` or ``(N, C, ...)`` with integer labels
+
+    Args:
+        num_classes: integer with number of classes. For binary, set to 1.
+        num_thresholds: number of bins used for computation. More bins will lead to more detailed
+            curve and accurate estimates, but will be slower and consume more memory. Default 100
+        compute_on_step:
+            Forward only calls ``update()`` and return None if this is set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step. default: False
+        process_group:
+            Specify the process group on which synchronization is called. default: None (which selects the entire world)
+
+    Example (binary case):
+        >>> from torchmetrics import BinnedPrecisionRecallCurve
+        >>> pred = torch.tensor([0, 0.1, 0.8, 0.4])
+        >>> target = torch.tensor([0, 1, 1, 0])
+        >>> pr_curve = BinnedPrecisionRecallCurve(num_classes=1, num_thresholds=5)
+        >>> precision, recall, thresholds = pr_curve(pred, target)
+        >>> precision
+        tensor([0.5000, 0.5000, 1.0000, 1.0000, 1.0000, 1.0000])
+        >>> recall
+        tensor([1.0000, 0.5000, 0.5000, 0.5000, 0.0000, 0.0000])
+        >>> thresholds
+        tensor([0.0000, 0.2500, 0.5000, 0.7500, 1.0000])
+
+    Example (multiclass case):
+        >>> pred = torch.tensor([[0.75, 0.05, 0.05, 0.05, 0.05],
+        ...                      [0.05, 0.75, 0.05, 0.05, 0.05],
+        ...                      [0.05, 0.05, 0.75, 0.05, 0.05],
+        ...                      [0.05, 0.05, 0.05, 0.75, 0.05]])
+        >>> target = torch.tensor([0, 1, 3, 2])
+        >>> pr_curve = BinnedPrecisionRecallCurve(num_classes=5, num_thresholds=3)
+        >>> precision, recall, thresholds = pr_curve(pred, target)
+        >>> precision   # doctest: +NORMALIZE_WHITESPACE
+        [tensor([0.2500, 1.0000, 1.0000, 1.0000]),
+        tensor([0.2500, 1.0000, 1.0000, 1.0000]),
+        tensor([2.5000e-01, 1.0000e-06, 1.0000e+00, 1.0000e+00]),
+        tensor([2.5000e-01, 1.0000e-06, 1.0000e+00, 1.0000e+00]),
+        tensor([2.5000e-07, 1.0000e+00, 1.0000e+00, 1.0000e+00])]
+        >>> recall   # doctest: +NORMALIZE_WHITESPACE
+        [tensor([1.0000, 1.0000, 0.0000, 0.0000]),
+        tensor([1.0000, 1.0000, 0.0000, 0.0000]),
+        tensor([1.0000, 0.0000, 0.0000, 0.0000]),
+        tensor([1.0000, 0.0000, 0.0000, 0.0000]),
+        tensor([0., 0., 0., 0.])]
+        >>> thresholds   # doctest: +NORMALIZE_WHITESPACE
+        [tensor([0.0000, 0.5000, 1.0000]),
+        tensor([0.0000, 0.5000, 1.0000]),
+        tensor([0.0000, 0.5000, 1.0000]),
+        tensor([0.0000, 0.5000, 1.0000]),
+        tensor([0.0000, 0.5000, 1.0000])]
+    """
+
+    TPs: Tensor
+    FPs: Tensor
+    FNs: Tensor
+    thresholds: Tensor
```

Comment:
is this a counter?
if I have one metric for train other for valid it would mix all together, right?

Reply:
This is just type annotation, it doesn't define a new variable and we use instance variables so there won't be overlap.

The cause here is the confusing way of Python type annotation for instance variables:
```python
class A:
   x: int # instance variable
   y: ClassVar[int] # class/static variable
```

https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations

Reply:
ok, I see not it is just kind or shadow to make IDE happy, right? https://github.com/PyTorchLightning/metrics/pull/128/files#r612738983
well, I find it confusing... :/
other thoughts? @PyTorchLightning/core-metrics

Reply:
Yeah, it's for type-checker, I do agree that given that we don't have that in other metrics anyway it's fine to remove from here as well.

Reply:
yes, so let's make this PR without it and then open separate PR just by introducing this typing for all metrics... ok?